File: src/agents/risk_manager.py
from langchain_core.messages import HumanMessage
from src.graph.state import AgentState, show_agent_reasoning
from src.utils.progress import progress
from src.tools.api import get_prices, prices_to_df
import json


##### Risk Management Agent #####
def risk_management_agent(state: AgentState):
    """Controls position sizing based on real-world risk factors for multiple tickers."""
    portfolio = state["data"]["portfolio"]
    data = state["data"]
    tickers = data["tickers"]

    # Initialize risk analysis for each ticker
    risk_analysis = {}
    current_prices_map = {}  # Store current prices for all relevant tickers

    # Step 1: Pre-fetch current prices for all tickers in the universe and in portfolio
    all_relevant_tickers = set(tickers) | set(portfolio.get("positions", {}).keys())
    for ticker_symbol in all_relevant_tickers:
        progress.update_status("risk_management_agent", ticker_symbol, "Fetching current price")
        # Fetch only the most recent price, effectively the end_date for this context
        # Assuming get_prices can fetch a single day if start_date and end_date are the same or close.
        # For simplicity, using data["end_date"] to get the latest known price.
        price_data_list = get_prices(
            ticker=ticker_symbol,
            start_date=data["end_date"], # Fetch for the "current" day
            end_date=data["end_date"],
        )
        if price_data_list:
            # Assuming the list is sorted or the last entry is the most recent if multiple are returned for a single day query
            current_prices_map[ticker_symbol] = price_data_list[-1].close 
        else:
            progress.update_status("risk_management_agent", ticker_symbol, "Warning: No current price data found")
            # Consider how to handle missing prices; for now, these tickers won't contribute to market value if missing

    # Step 2: Calculate Net Liquidation Value (NLV)
    progress.update_status("risk_management_agent", None, "Calculating Net Liquidation Value")
    total_portfolio_value = portfolio.get("cash", 0.0)

    for ticker_symbol, position_details in portfolio.get("positions", {}).items():
        market_price = current_prices_map.get(ticker_symbol)
        if market_price is not None:
            long_shares = position_details.get("long", 0)
            short_shares = position_details.get("short", 0)
            total_portfolio_value += long_shares * market_price
            total_portfolio_value -= short_shares * market_price # Subtract market value of short positions
        else:
            # If price is missing for a held ticker, its market value contribution is skipped.
            # Alternative: could use cost basis, but NLV usually requires market prices.
            progress.update_status("risk_management_agent", ticker_symbol, f"Warning: Price unavailable for held ticker {ticker_symbol}, not included in NLV market value calculation.")


    # Step 3: Calculate position limits for each ticker in the trading universe
    progress.update_status("risk_management_agent", None, "Calculating position limits per ticker")
    for ticker in tickers: # Iterate through the universe of tickers for this run
        current_price = current_prices_map.get(ticker)
        if current_price is None:
            progress.update_status("risk_management_agent", ticker, "Failed: No current price data, cannot calculate limits.")
            risk_analysis[ticker] = { # Provide a default/error state for this ticker
                "remaining_position_limit": 0.0,
                "current_price": 0.0,
                "reasoning": {
                    "portfolio_value": float(total_portfolio_value),
                    "error": "Missing current price for ticker, limits cannot be determined."
                }
            }
            continue

        # Calculate current market value of existing position for this specific ticker
        existing_position_details = portfolio.get("positions", {}).get(ticker, {})
        long_value = existing_position_details.get("long", 0) * current_price
        short_value = existing_position_details.get("short", 0) * current_price
        # For position limit calculations, often the gross market value of the specific position is considered.
        # Or, more simply, how much more of this specific ticker can be added.
        # Let's use the absolute value for "current_position_value_for_limit_calc"
        current_position_value_for_limit_calc = abs(long_value - short_value)


        # Base limit is 20% of total_portfolio_value for any single position (absolute exposure)
        single_ticker_max_exposure = total_portfolio_value * 0.20

        # How much more exposure can be added to this ticker
        remaining_exposure_allowance = single_ticker_max_exposure - current_position_value_for_limit_calc
        
        # Convert this remaining exposure allowance to a cash amount for new positions
        # This is the cash value that can be used to establish or increase a position.
        cash_for_new_position = max(0, remaining_exposure_allowance)

        # Ensure this doesn't exceed available cash for new buys/shorts (considering margin for shorts)
        # This part can be complex depending on how margin is handled for new short positions.
        # For simplicity, let's assume for long positions, it's capped by available cash.
        # And for short positions, it's also notionally capped by this amount before margin considerations.
        max_cash_for_ticker_position = min(cash_for_new_position, portfolio.get("cash", 0.0))


        risk_analysis[ticker] = {
            "remaining_position_limit": float(max_cash_for_ticker_position), # This is the cash value limit for new trades on this ticker
            "current_price": float(current_price),
            "reasoning": {
                "total_portfolio_nlv": float(total_portfolio_value),
                "single_ticker_max_exposure_pct": 0.20,
                "single_ticker_max_exposure_value": float(single_ticker_max_exposure),
                "current_ticker_market_value": float(current_position_value_for_limit_calc),
                "remaining_exposure_allowance_for_ticker": float(remaining_exposure_allowance),
                "cash_available_for_new_trades_on_ticker": float(max_cash_for_ticker_position),
                "total_portfolio_cash": float(portfolio.get("cash", 0.0)),
            },
        }
        progress.update_status("risk_management_agent", ticker, "Done")

    message = HumanMessage(
        content=json.dumps(risk_analysis),
        name="risk_management_agent",
    )

    if state["metadata"]["show_reasoning"]:
        show_agent_reasoning(risk_analysis, "Risk Management Agent")

    # Add the signal to the analyst_signals list
    state["data"]["analyst_signals"]["risk_management_agent"] = risk_analysis

    return {
        "messages": state["messages"] + [message],
        "data": data,
    }

File: src/agents/risk_manager.rs
// Source: src/agents/risk_manager.py
//! Sibling to src/agents/risk_manager.py
//! Manages portfolio risk by calculating position sizing limits adjusted for volatility and asset correlation.

use anyhow::{Result, Context};
use std::collections::{HashMap, HashSet};
use crate::graph::state::AgentState;
use crate::tools::api::get_prices;

/// Volatility adjusted risk management agent entry node.
pub async fn risk_management_agent(state: &mut AgentState, agent_id: &str) -> Result<()> {
    println!("Running Risk Management Agent: {}", agent_id);

    let start_date = state.data.get("start_date")
        .and_then(|v| v.as_str())
        .context("Missing start_date in state data")?;

    let end_date = state.data.get("end_date")
        .and_then(|v| v.as_str())
        .context("Missing end_date in state data")?;

    let tickers_json = state.data.get("tickers")
        .context("Missing tickers in state data")?;
    let tickers: Vec<String> = serde_json::from_value(tickers_json.clone())?;

    let portfolio = state.data.get("portfolio")
        .context("Missing portfolio in state data")?;

    let api_key = state.metadata.get("FINANCIAL_DATASETS_API_KEY")
        .and_then(|v| v.as_str());

    let mut risk_analysis = serde_json::Map::new();
    let mut current_prices = HashMap::new();
    let mut annualized_vols = HashMap::new();
    let mut returns_by_ticker = HashMap::new();

    // Determine all active positions
    let positions_map = portfolio.get("positions")
        .and_then(|p| p.as_object());
    
    let mut all_tickers = HashSet::new();
    for t in &tickers {
        all_tickers.insert(t.clone());
    }
    if let Some(pos_obj) = positions_map {
        for k in pos_obj.keys() {
            all_tickers.insert(k.clone());
        }
    }

    for ticker in &all_tickers {
        let prices = match get_prices(ticker, start_date, end_date, api_key).await {
            Ok(p) => p,
            Err(_) => {
                current_prices.insert(ticker.clone(), 0.0);
                annualized_vols.insert(ticker.clone(), 0.25);
                continue;
            }
        };

        if prices.len() < 2 {
            current_prices.insert(ticker.clone(), 0.0);
            annualized_vols.insert(ticker.clone(), 0.25);
            continue;
        }

        let closes: Vec<f64> = prices.iter().map(|p| p.close).collect();
        let last_price = closes[closes.len() - 1];
        current_prices.insert(ticker.clone(), last_price);

        // Daily returns volatility
        let mut daily_returns = Vec::new();
        for i in 1..closes.len() {
            let prev = closes[i - 1];
            daily_returns.push((closes[i] - prev) / if prev == 0.0 { 1e-8 } else { prev });
        }

        let mean = daily_returns.iter().sum::<f64>() / daily_returns.len() as f64;
        let variance = daily_returns.iter().map(|&x| (x - mean).powi(2)).sum::<f64>() / daily_returns.len() as f64;
        let daily_vol = variance.sqrt();
        let ann_vol = daily_vol * 252.0_f64.sqrt();

        annualized_vols.insert(ticker.clone(), ann_vol);
        returns_by_ticker.insert(ticker.clone(), daily_returns);
    }

    // Correlation multiplier calculation
    let mut avg_correlations = HashMap::new();
    for t1 in &tickers {
        let mut correlations = Vec::new();
        if let Some(r1) = returns_by_ticker.get(t1) {
            for t2 in &all_tickers {
                if t1 == t2 {
                    continue;
                }
                if let Some(r2) = returns_by_ticker.get(t2) {
                    let len = std::cmp::min(r1.len(), r2.len());
                    if len > 5 {
                        let slice1 = &r1[r1.len() - len..];
                        let slice2 = &r2[r2.len() - len..];
                        let mean1 = slice1.iter().sum::<f64>() / len as f64;
                        let mean2 = slice2.iter().sum::<f64>() / len as f64;
                        let cov = slice1.iter().zip(slice2.iter()).map(|(&x, &y)| (x - mean1) * (y - mean2)).sum::<f64>() / len as f64;
                        let var1 = slice1.iter().map(|&x| (x - mean1).powi(2)).sum::<f64>() / len as f64;
                        let var2 = slice2.iter().map(|&x| (x - mean2).powi(2)).sum::<f64>() / len as f64;
                        let std1 = var1.sqrt();
                        let std2 = var2.sqrt();
                        let corr = if std1 > 0.0 && std2 > 0.0 { cov / (std1 * std2) } else { 0.0 };
                        correlations.push(corr);
                    }
                }
            }
        }
        let avg_corr = if correlations.is_empty() {
            0.0
        } else {
            correlations.iter().sum::<f64>() / correlations.len() as f64
        };
        avg_correlations.insert(t1.clone(), avg_corr);
    }

    // Portfolio Value Points
    let cash = portfolio.get("cash").and_then(|v| v.as_f64()).unwrap_or(100000.0);
    let mut total_portfolio_value = cash;
    if let Some(pos_obj) = positions_map {
        for (ticker, pos_val) in pos_obj {
            if let Some(price) = current_prices.get(ticker) {
                let long = pos_val.get("long").and_then(|v| v.as_i64()).unwrap_or(0) as f64;
                let short = pos_val.get("short").and_then(|v| v.as_i64()).unwrap_or(0) as f64;
                total_portfolio_value += long * price;
                total_portfolio_value -= short * price;
            }
        }
    }

    // Calculate limit per ticker
    for ticker in &tickers {
        let current_price = current_prices.get(ticker).copied().unwrap_or(0.0);
        if current_price <= 0.0 {
            risk_analysis.insert(
                ticker.clone(),
                serde_json::json!({
                    "remaining_position_limit": 0.0,
                    "current_price": 0.0,
                    "reasoning": {
                        "error": "Missing price data for risk calculation"
                    }
                })
            );
            continue;
        }

        let ann_vol = annualized_vols.get(ticker).copied().unwrap_or(0.25);
        let vol_limit_pct = calculate_volatility_adjusted_limit(ann_vol);

        let avg_corr = avg_correlations.get(ticker).copied().unwrap_or(0.0);
        let corr_mult = calculate_correlation_multiplier(avg_corr);

        let combined_limit_pct = vol_limit_pct * corr_mult;
        let position_limit = total_portfolio_value * combined_limit_pct;

        // Current position absolute exposure
        let mut current_exposure = 0.0;
        if let Some(pos_obj) = positions_map {
            if let Some(pos_val) = pos_obj.get(ticker) {
                let long = pos_val.get("long").and_then(|v| v.as_i64()).unwrap_or(0) as f64;
                let short = pos_val.get("short").and_then(|v| v.as_i64()).unwrap_or(0) as f64;
                current_exposure = (long * current_price - short * current_price).abs();
            }
        }

        let remaining_limit = (position_limit - current_exposure).max(0.0);
        let max_position_size = remaining_limit.min(cash);

        risk_analysis.insert(
            ticker.clone(),
            serde_json::json!({
                "remaining_position_limit": max_position_size,
                "current_price": current_price,
                "volatility_metrics": {
                    "annualized_volatility": ann_vol
                },
                "correlation_metrics": {
                    "avg_correlation_with_active": avg_corr
                },
                "reasoning": {
                    "portfolio_value": total_portfolio_value,
                    "current_position_value": current_exposure,
                    "base_position_limit_pct": vol_limit_pct,
                    "correlation_multiplier": corr_mult,
                    "combined_position_limit_pct": combined_limit_pct,
                    "position_limit": position_limit,
                    "remaining_limit": remaining_limit,
                    "available_cash": cash
                }
            })
        );
    }

    let analyst_signals = state.data.entry("analyst_signals".to_string())
        .or_insert_with(|| serde_json::json!({}));
    
    if let Some(obj) = analyst_signals.as_object_mut() {
        obj.insert(agent_id.to_string(), serde_json::Value::Object(risk_analysis));
    }

    Ok(())
}

pub fn calculate_volatility_adjusted_limit(annualized_volatility: f64) -> f64 {
    let base_limit = 0.20;
    let vol_multiplier = if annualized_volatility < 0.15 {
        1.25
    } else if annualized_volatility < 0.30 {
        1.0 - (annualized_volatility - 0.15) * 0.5
    } else if annualized_volatility < 0.50 {
        0.75 - (annualized_volatility - 0.30) * 0.5
    } else {
        0.50
    };
    let vol_multiplier = vol_multiplier.max(0.25).min(1.25);
    base_limit * vol_multiplier
}

pub fn calculate_correlation_multiplier(avg_correlation: f64) -> f64 {
    if avg_correlation >= 0.80 {
        0.70
    } else if avg_correlation >= 0.60 {
        0.85
    } else if avg_correlation >= 0.40 {
        1.00
    } else if avg_correlation >= 0.20 {
        1.05
    } else {
        1.10
    }
}
